### PR TITLE
Move blas and math_function from operators/math to framework.

### PR DESCRIPTION
**Pull request by wanglei828**
@wangkuiyi The PR will contains following changes:

I moved the following files from operators/math to framework:
paddle/fluid/framework/blas.cc
paddle/fluid/framework/blas.h
paddle/fluid/framework/blas_impl.cu.h
paddle/fluid/framework/blas_impl.h
paddle/fluid/framework/math_function.cc
paddle/fluid/framework/math_function.cu
paddle/fluid/framework/math_function.h
paddle/fluid/framework/math_function_impl.h
paddle/fluid/framework/math_function_test.cc
paddle/fluid/framework/math_function_test.cu

Also I added header files from original paddle/fluid/framework because they are included in blas.h file.
paddle/fluid/framework/block_desc.h
paddle/fluid/framework/library_type.h
paddle/fluid/framework/op_desc.h
paddle/fluid/framework/op_info.h
paddle/fluid/framework/op_kernel_type.h
paddle/fluid/framework/operator.h
paddle/fluid/framework/proto_desc.h
paddle/fluid/framework/type_defs.h
paddle/fluid/framework/var_desc.h

Considering you have changed the "data_layout.h" file in commit #20 , the type changed from "enum class" to "class".
In the "paddle/fluid/framework/op_kernel_type.h" line 33:
int data_layout = static_cast<int>(key.data_layout) << (LEFT_SHIFT * 2); 
cause compilation error.
I changed to:
int data_layout = static_cast<int>(key.data_layout_.type_) << (LEFT_SHIFT * 2);

### CODE REVIEW

File: cmake/cblas.cmake [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# Find the CBlas and lapack libraries
```
Comment (wangkuiyi) [inline cmake/cblas.cmake]: 我看了一下这个文件的源码，按照CMake的标准包的naming convention，比如

- FindCUDA https://cmake.org/cmake/help/v3.0/module/FindCUDA.html
- FindBLAS https://cmake.org/cmake/help/v3.0/module/FindBLAS.html

这个文件应该叫做 FindCBLAS.cmake 。

一般第三方写的CMake插件也都按照这个naming convention，比如

- Caffe 的 FindVecLib.cmake https://github.com/BVLC/caffe/blob/master/cmake/Modules/FindvecLib.cmake
- Caffe2 的 FindMKL: https://github.com/caffe2/caffe2/blob/master/cmake/Modules/FindMKL.cmake
- MxNet 的 FindMKL： https://github.com/apache/incubator-mxnet/blob/master/cmake/Modules/FindMKL.cmake

File: cmake/cblas.cmake [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+if(WITH_MKLML AND MKLML_INC_DIR AND MKLML_LIB)
+  set(CBLAS_FOUND ON)
+  set(CBLAS_PROVIDER MKLML)
+  set(CBLAS_INC_DIR ${MKLML_INC_DIR})
+  set(CBLAS_LIBRARIES ${MKLML_LIB})
+
+  add_definitions(-DPADDLE_WITH_MKLML)
+  add_definitions(-DLAPACK_FOUND)
+
+  message(STATUS "Found cblas and lapack in MKLML "
```
Comment (wangkuiyi) [inline cmake/cblas.cmake]: cblas => CBLAS
lapack => LAPACK

File: cmake/cblas.cmake [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+# If any cblas implementation found, the following variable will be set.
+#    CBLAS_PROVIDER  # one of MKLML, OPENBLAS, REFERENCE
+#    CBLAS_INC_DIR   # the include directory for cblas.
+#    CBLAS_LIBS      # a list of libraries should be linked by paddle.
+#                    # Each library should be full path to object file.
+
+set(CBLAS_FOUND OFF)
+
+## Find MKLML First.
+if(WITH_MKLML AND MKLML_INC_DIR AND MKLML_LIB)
```
Comment (wangkuiyi) [inline cmake/cblas.cmake]: 这里的意思是用户需要手动指定 MKLML_INC_DIR 和 MKLML_LIB ? 我感觉我们可以抄袭一下MxNet的[FindMKL.cmake](https://github.com/apache/incubator-mxnet/blob/master/cmake/Modules/FindMKL.cmake)，这个插件可以搜索 MKL、MKLDNN、MKLML。然后在这里调用 FindMKL.cmake。

一个例子供你参考：[这个find_cudnn.cmake](https://github.com/wangkuiyi/fluid/blob/develop/cmake/find_cudnn.cmake) 是我抄袭的Caffe2 的。

我说的抄袭，是真个文件拷贝过来，最多只是稍作修改；而不是说抄袭对方文件中的一段代码 —— 这样很难trace从而抄袭来的，不利于长期维护源码。

File: cmake/cblas.cmake [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+  add_definitions(-DPADDLE_WITH_MKLML)
+  add_definitions(-DLAPACK_FOUND)
+
+  message(STATUS "Found cblas and lapack in MKLML "
+    "(include: ${CBLAS_INC_DIR}, library: ${CBLAS_LIBRARIES})")
+  return()
+endif()
+
+## Then find openblas.
```
Comment (wangkuiyi) [inline cmake/cblas.cmake]: 这段和 OpenBLAS 相关的代码是从 https://github.com/BVLC/caffe/blob/master/cmake/Modules/FindOpenBLAS.cmake 拷贝粘贴的。我建议我们要抄袭就真个拷贝被人的文件 —— 这样利于trace抄袭来源，其实也好给原作者credit —— 然后我们的cmake调用这个 FindOpenBLAS.cmake —— 这样，我们自己的 FindCBLAS.cmake 也不会像现在这样冗长。

File: cmake/cblas.cmake [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  add_definitions(-DPADDLE_USE_OPENBLAS)
+  add_definitions(-DLAPACK_FOUND)
+
+  message(STATUS "Found OpenBLAS (include: ${OPENBLAS_INC_DIR}, library: ${CBLAS_LIBRARIES})")
+  message(STATUS "Found lapack in OpenBLAS (include: ${OPENBLAS_LAPACKE_INC_DIR})")
+  return()
+endif()
+
+
+## Then find the reference-cblas.  www.netlib.org/blas/
```
Comment (wangkuiyi) [inline cmake/cblas.cmake]: 之类有个英语语言错误 —— “参考实现” 是 reference implementation，而不是 reference。这里应该写： 

> Then, try to find the reference implementation of CBLAS: www.netlib.org/blas.

File: cmake/cblas.cmake [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+set(REFERENCE_CBLAS_ROOT $ENV{REFERENCE_CBLAS_ROOT} CACHE PATH
```
Comment (wangkuiyi) [inline cmake/cblas.cmake]: REFERENCE_CBLAS_ROOT => NETLIB_CBLAS_ROOT

**Comment by tonyyang-svail** [comment]
math_function and blas are in the develop branch now.

**Pull request closed without merging**